NLB-7031: Update Native OIDC process in NginxaaS docs (#1176)
nginx/documentation: content/nginxaas-azure/quickstart/security-controls/oidc.md

Learn how to configure F5 NGINXaaS for Azure with OpenID Connect (OIDC) authentication.

There are currently two methods available for setting up OIDC authentication.

1. Using Native OIDC implementation (Introduced from NGINX Plus R35)
   

>    This method applies to NGINX Plus Release 35 and later. In earlier versions, NGINX Plus relied on an njs-based solution, which required NGINX JavaScript files, key-value stores, and advanced OpenID Connect logic. In the latest NGINX Plus version, the new [OpenID Connect module](https://nginx.org/en/docs/http/ngx_http_oidc_module.html) simplifies this process to just a few directives.

Reviewer: This paragraph starts with "this method" but is followed by prerequisites that apply to both methods.

Maybe it's better to move this paragraph to Line 33 after the Native OIDC heading.

2. Using NJS based implementation

## Prerequisites

These prerequisites are used for both methods of configuring NGINXaaS for Azure with IdP using Native OIDC and NJS.
1. Configure an NGINXaaS deployment with [SSL/TLS certificates]({{< ref "/nginxaas-azure/getting-started/ssl-tls-certificates/" >}}).

2. Enable [Runtime State Sharing]({{< ref "/nginxaas-azure/quickstart/runtime-state-sharing.md" >}}) on the NGINXaaS deployment.

## Configure NGINXaaS for Azure with IdP using Native OIDC

> ### Prerequisites
> 1. Configure the IdP. For example, you can [register a Microsoft Entra Web application]({{< ref "/nginx/deployment-guides/single-sign-on/entra-id/#entra-setup" >}}) as the IdP.
> 2. A domain name pointing to your NGINXaaS deployment, for example, `demo.example.com`. This will be referred to as `<nginxaas_deployment_fqdn>` throughout this guide.

Reviewer: ```suggestion
### Prerequisites

1. Configure the IdP. For example, you can [register a Microsoft Entra Web application]({{< ref "/nginx/deployment-guides/single-sign-on/entra-id/#entra-setup" >}}) as the IdP.
1. A domain name pointing to your NGINXaaS deployment, for example, `demo.example.com`. This will be referred to as `<nginxaas_deployment_fqdn>` throughout this guide.
```

This is currently not enforced but 2 markdown good practices:
- Leave a blank line after headings
- Numbered list can be a series of 1. 1. 1. ; Markdown will number these appropriately if the correct indenting is respected. It makes easier to add or remove steps later on.

Author: Addressed blank line after heading and Numbered list series

With your IdP configured, you can enable OIDC on NGINXaaS for Azure.

1. Ensure that you have the values of the **Client ID**, **Client Secret**, and **Tenant ID** obtained during IdP configuration.

> 2. In your NGINX configuration file, add a public DNS resolver with the [`resolver`](https://nginx.org/en/docs/http/ngx_http_core_module.html#resolver) directive in the [`http {}`](https://nginx.org/en/docs/http/ngx_http_core_module.html#http) context:

Reviewer: ```suggestion
1. In your NGINX configuration file, add a public DNS resolver with the [`resolver`](https://nginx.org/en/docs/http/ngx_http_core_module.html#resolver) directive in the [`http {}`](https://nginx.org/en/docs/http/ngx_http_core_module.html#http) context:
```

        oidc_provider entra {
            issuer            https://login.microsoftonline.com/<tenant_id>/v2.0;
            client_id         <client_id>;
            client_secret     <client_secret>;
            session_store      my_store;
            logout_uri        /logout;
            post_logout_uri   https://<nginxaas_deployment_fqdn>/post_logout/;
            logout_token_hint on;
            userinfo          on;
        }
        # ...
    }
    ```

    Where:
    - `<tenant_id>` is your Microsoft Entra Tenant ID
    - `<client_id>` is your Application (client) ID from Entra ID
    - `<client_secret>` is your client secret from Entra ID
    - `<nginxaas_deployment_fqdn>` is your NGINXaaS deployment FQDN

    {{< call-out "note" >}} The `state=/opt/oidc_sessions.json` parameter enables persistence of OIDC session data across NGINX restarts. The state file path must be placed in a directory accessible to the NGINX worker processes, following [NGINX Filesystem Restrictions]({{< ref "/nginxaas-azure/getting-started/nginx-configuration/overview/#nginx-filesystem-restrictions" >}}).{{< /call-out >}}

> 4. Configure your server block with OIDC protection. This example uses localhost as the upstream server:

Reviewer: ```suggestion
1. Configure your server block with OIDC protection. The following example uses localhost as the upstream server:
```

    ```nginx
    server {
        listen 443 ssl;
        server_name <nginxaas_deployment_fqdn>;

        ssl_certificate /etc/ssl/certs/fullchain.pem;
        ssl_certificate_key /etc/ssl/private/key.pem;

        location / {
            # Protect this location with OIDC
            auth_oidc entra;

            # Forward OIDC claims as headers
            proxy_set_header sub   $oidc_claim_sub;
            proxy_set_header email $oidc_claim_email;
            proxy_set_header name  $oidc_claim_name;

            proxy_pass http://127.0.0.1:8080;
        }

        location /post_logout/ {
            return 200 "You have been logged out.\n";
            default_type text/plain;
        }
    }
    
    server {
        # Simple test upstream server
        listen 8080;
        location / {
            return 200 "Hello, $http_name!\nEmail: $http_email\nEntra ID sub: $http_sub\n";
            default_type text/plain;
        }
    }
    ```

> 5. Add the runtime state sharing configuration to your NGINX configuration as mentioned in the [Prerequisites](#prerequisites). This enables synchronization of OIDC session data across NGINXaaS instances:

Reviewer: ```suggestion
1. Add the runtime state sharing configuration to your NGINX configuration as mentioned in the [Prerequisites]({{< ref "/nginxaas-azure/quickstart/security-controls/oidc.md#prerequisites" >}}). This enables synchronization of OIDC session data across NGINXaaS instances:
```

Author: Addressed the numbering series and adding ref as suggested above.

                auth_oidc entra;

                # Forward OIDC claims as headers if desired
                proxy_set_header sub   $oidc_claim_sub;
                proxy_set_header email $oidc_claim_email;
                proxy_set_header name  $oidc_claim_name;

                proxy_pass http://127.0.0.1:8080;
            }

            location /post_logout/ {
                return 200 "You have been logged out.\n";
                default_type text/plain;
            }
        }

        server {
            # Simple test upstream server
            listen 8080;

            location / {
                return 200 "Hello, $http_name!\nEmail: $http_email\nEntra ID sub: $http_sub\n";
                default_type text/plain;
            }
        }
    }

    stream {
        resolver 127.0.0.1:49153 valid=20s;

        server {
            listen 9000;
            zone_sync;
            zone_sync_server internal.nginxaas.nginx.com:9000 resolve;
        }
    }
    ```
    </details>

6. Upload the NGINX configurations. See [Upload an NGINX configuration]({{< ref "/nginxaas-azure/getting-started/nginx-configuration/" >}}) for more details.

> For more detailed steps on this OIDC configuration, please refer to:

Author: Addressed the numbering series and spacing as sugessted above.

- [Single Sign-On with Microsoft Entra ID]({{< ref "/nginx/deployment-guides/single-sign-on/entra-id.md" >}})
- [Terraform snippets for Native OIDC use case](https://github.com/nginxinc/nginxaas-for-azure-snippets/tree/main/terraform/configurations/native-oidc)

> ### Testing

Reviewer: ```suggestion

### Testing
```

1. Open `https://<nginxaas_deployment_fqdn>/` in a browser. You will be automatically redirected to your IdP sign-in page.

2. Enter valid IdP credentials. Upon successful sign-in, you will be redirected back to NGINXaaS and see your protected application. Using the example configuration, you will see a message displaying the authenticated user's information in the browser:

    ```text
    Hello, [Name]!
    Email: [email]
    Entra ID sub: [subject_id]
    ```

3. To test logout, navigate to `https://<nginxaas_deployment_fqdn>/logout`. NGINXaaS initiates an RP-initiated logout, and your IdP ends the session and redirects back to the post-logout page.

## Configure NGINXaaS for Azure with IdP using NJS
### Prerequisites

> 1. [Configure the IdP](https://github.com/nginxinc/nginx-openid-connect/blob/main/README.md#configuring-your-idp). For example, you can [register a Microsoft Entra Web application](https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app) as the IdP.

Author: Addressed the spacing as suggested above.

2. Enable [Runtime State Sharing]({{< ref "/nginxaas-azure/quickstart/runtime-state-sharing.md" >}}) on the NGINXaaS deployment.

These prerequisites are used for both methods of configuring NGINXaaS for Azure with IdP using Native OIDC and NJS.

There are currently two methods available for setting up OIDC authentication.

> 1. Using Native OIDC implementation (Introduced from NGINX Plus R35)

Author: correct feature implemetation of native oidc took place in R35, I took reference of https://docs.nginx.com/nginx/deployment-guides/single-sign-on/entra-id/. Thats the reason I added R35

> 1. Using Native OIDC implementation (Introduced from NGINX Plus R35)

Reviewer 2: No. The feature itself appeared in R34 (aka native oidc implementation). In R35, 2 more enhancements to oidcs were added: user-initiated logout and userinfo endpoint features were added. Pls see the changelog: https://docs.nginx.com/nginx/releases/
So, if you're talking about native OIDC as a new feature, it's R34, if you're talking about OIDC enhancements, (2 of them) - it's R35.

        location / {
            # Protect this location with OIDC
            auth_oidc entra;

            # Forward OIDC claims as headers
            proxy_set_header sub   $oidc_claim_sub;
            proxy_set_header email $oidc_claim_email;
            proxy_set_header name  $oidc_claim_name;

            proxy_pass http://127.0.0.1:8080;
        }

        location /post_logout/ {
            return 200 "You have been logged out.\n";
            default_type text/plain;
        }
    }
    
    server {
        # Simple test upstream server
        listen 8080;
        location / {
            return 200 "Hello, $http_name!\nEmail: $http_email\nEntra ID sub: $http_sub\n";
            default_type text/plain;
        }
    }
    ```

5. Add the runtime state sharing configuration to your NGINX configuration as mentioned in the [Prerequisites](#prerequisites). This enables synchronization of OIDC session data across NGINXaaS instances:

    ```nginx
    stream {
        resolver 127.0.0.1:49153 valid=20s;

        server {
            listen 9000;
            zone_sync;
            zone_sync_server internal.nginxaas.nginx.com:9000 resolve;
        }
    }
    ```

>     <details close>

Reviewer 3: We recently updated our hugo version, and now support the native hugo [details shortcode](https://gohugo.io/shortcodes/details/). 
We'll do a refactor pass soon, to replace all usages with this version.
